giskard_vision/core/detectors/metadata_scan_detector.py

```diff
@@ -0,0 +1,171 @@
+        df_for_scan = self.get_df_for_scan(model, dataset, list_metadata)
+
+        df_for_prediction = df_for_scan.copy()
+
+        def prediction_function(df: pd.DataFrame) -> np.ndarray:
+            return pd.merge(df, df_for_prediction, on="index", how="inner")["prediction"].values
+
+        # Create Giskard dataset and model
+        giskard_dataset = Dataset(df=df_for_scan.copy(), target="target", cat_columns=list_categories + ["index"])
+
+        giskard_model = Model(
+            model=prediction_function,
+            model_type=self.type_task,
+            feature_names=list_metadata + ["index"],
+        )
+
+        # Get scan results
+        results = scan(giskard_model, giskard_dataset, max_issues_per_detector=None, verbose=False)
+
+        list_scan_results = []
+
+        # For each slice found, get appropriate scna results with the metric
+        for issue in results.issues:
+            current_data_slice = giskard_dataset.slice(issue.slicing_fn)
+            indices = list(current_data_slice.df.sort_values(by="metric", ascending=False)["index"].values)
+            filenames = (
+                [dataset.get_image_path(idx) for idx in indices[: self.num_images]]
+                if hasattr(dataset, "get_image_path")
+                else []
+            )
+            list_scan_results.append(
+                self.get_scan_result(
+                    metric_value=current_data_slice.df["metric"].mean(),
+                    metric_reference_value=giskard_dataset.df["metric"].mean(),
+                    metric_name=self.metric.name,
+                    filename_examples=filenames,
+                    name=issue.slicing_fn.meta.display_name,
+                    size_data=len(current_data_slice.df),
+                    issue_group=meta.issue_group(issue.features[0]),
+                )
+            )
+
+        return list_scan_results
+
+    def get_df_for_scan(self, model: Any, dataset: Any, list_metadata: Sequence[str]) -> pd.DataFrame:
+        # Create a dataframe containing each metadata and metric, surrogate target, surrogate prediction
+        # image path for display in html, and index
+        df = {name_metadata: [] for name_metadata in list_metadata}
+        df["metric"] = []
+        df["target"] = []
+        df["prediction"] = []
+        df["index"] = []
+
+        # For now the DataFrame is built without a batch strategy because
+        # we need the metadata, labels and image path on an individual basis,
+        # and sometimes the model may fail on an image.
+        # TODO: make this cleaner and more efficient with batch computations
+
+        for i in range(len(dataset)):
+            try:
+                image = dataset.get_image(i)
+                prediction_result = model.predict_image(image)
+                ground_truth = dataset.get_labels(i)
+                metadata = dataset.get_meta(i)
+                metric_value = self.metric.get(prediction_result, ground_truth)
```

Comment:
you'd only need to wrap the prediction in a PredictionResult class, in order not to break the MetricBase API here